cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,362 @@
+)
+
+var propOrder = map[string]int{
+	typeKey:        1,
+	formatKey:      2,
+	descriptionKey: 3,
+	itemsKey:       4,
+	propertiesKey:  5,
+	defaultKey:     6,
+}
+
+const (
+	objectVal = "object"
+	arrayVal  = "array"
+	floatVal  = "float"
+)
+
+// Yaml tags that differentiate the type of scalar object in the node
+const (
+	nullTag  = "!!null"
+	boolTag  = "!!bool"
+	intTag   = "!!int"
+	floatTag = "!!float"
+)
+
+type MapItem struct {
+	Key   interface{}
+	Value interface{}
+}
+
+type Map struct {
+	Items []*MapItem
+}
+
+func (m *Map) Iterate(iterFunc func(k, v interface{})) {
+	for _, item := range m.Items {
+		iterFunc(item.Key, item.Value)
+	}
+}
+
+type openAPIKeys []*MapItem
+
+func (o openAPIKeys) Len() int {
+	return len(o)
+}
+
+func (o openAPIKeys) Less(i, j int) bool {
+	return propOrder[o[i].Key.(string)] < propOrder[o[j].Key.(string)]
+}
+
+func (o openAPIKeys) Swap(i, j int) {
+	o[i], o[j] = o[j], o[i]
+}
+
+type HelmValuesSchemaGen struct {
+	// dirPath to the helm chart directory
+	dirPath string
+}
+
+func NewHelmValuesSchemaGen(path string) HelmValuesSchemaGen {
+	return HelmValuesSchemaGen{path}
+}
+
+func (h HelmValuesSchemaGen) Schema() (*kcdatav1alpha1.ValuesSchema, error) {
+	fileData, err := h.readValuesFile()
+	if err != nil {
+		return nil, err
+	}
+	if len(fileData) == 0 {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+
+	var document yaml3.Node
+	err = yaml3.Unmarshal(fileData, &document)
+	if err != nil {
+		return nil, err
+	}
+	if document.Kind == 0 {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+	if document.Kind != yaml3.DocumentNode {
+		// return proper error message
+		return nil, fmt.Errorf("invalid node kind supplied: %d", document.Kind)
+	}
+	if document.Content[0].Kind != yaml3.MappingNode {
+		return nil, fmt.Errorf("values file must resolve to a map (was %d)", document.Content[0].Kind)
+	}
+	openAPIProperties, err := h.calculateProperties(nil, document.Content[0])
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := yaml2.Marshal(h.convertToYAML(openAPIProperties))
+	if err != nil {
+		return nil, err
+	}
+	jsonEncodedBytes, err := yaml.YAMLToJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return &kcdatav1alpha1.ValuesSchema{
+		OpenAPIv3: runtime.RawExtension{Raw: jsonEncodedBytes},
+	}, nil
+}
+
+func (h HelmValuesSchemaGen) readValuesFile() ([]byte, error) {
+	fileInfo, err := os.Stat(h.dirPath)
+	if err != nil {
+		return nil, err
+	}
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("expected %s to be directory", h.dirPath)
+	}
+
+	fileData, err := os.ReadFile(filepath.Join(h.dirPath, "values.yaml"))
+	if err != nil {
+		// It is possible that helm chart doesn't have values.yml file.
+		if os.IsNotExist(err){
+			return nil, nil
+		}
+		return nil, err
+	}
+	return fileData, nil
+}
+
+func (h HelmValuesSchemaGen) convertToYAML(val interface{}) interface{} {
```

Comment:
Are we sure this covers all the cases? 
Are we ensuring in `calculateProperties` that we won't end up with slices as values here? Does this still cover the case?

I think using a function exported by the YAML library would be safer?

Reply:
> Are we ensuring in calculateProperties that we won't end up with slices as values here? Does this still cover the case?

Yes, all the slices values become properties. We cant use the library function as it is because that adds `key` and `value` literal while marshalling.

Reply:
IMHO as it takes interface{} as an input, thats why I have named it generic. If you still think `mapToYAMLBytes` sounds better, I can make the change

Reply:
> IMHO as it takes interface{} as an input, thats why I have named it generic

+1

How about `toYAML()`?

Reply:
Adopted